### PR TITLE
Support for Call Hierarchy for LSP

### PR DESCRIPTION
**Pull request by sdedic**
This PR adds support for `textDocument/prepareCallHierarchy`, `callHierarchy/incomingCalls` and `callHierarchy/outgoingCalls` messages (see [LSP Protocol Spec 3.16](https://microsoft.github.io/language-server-protocol/specification#textDocument_prepareCallHierarchy) ).

I decided to **reuse** `StructureElement` invented as a part of Outline support: elements reported by the Call Hierarchy have exactly the same meaning, plus the `uri`, which is implicit in the Outline. I have added `file` property to `StructureElement` and relevant `StructureProvider.Builder` method.

In order to share code between Java Editor module (that implements Outline) and Java Refactoring module (which implements the Call Hierarchy, but does not export it through any API), I've introduced helper methods for LSP into `java.sourceui` module: `lsp.api` module represents an abstract UI, after all, so adapting Java-specific structures to abstract ones seemed reasonable. This API allows to move Java Element -> StructureElement conversion to `java.sourceui`, and can be called from both `java.editor` and `java.refactoring`.

The other reason to choose `java.sourceui` for introducing the API is that the `ElementOpen` implementation is here, and can be adjusted to support StructureElement creation without exposing additional APIs.

### CODE REVIEW

File: java/refactoring.java/nbproject/project.xml [hunk 1]
```diff
@@ -52,6 +52,15 @@
                         <specification-version>1.0</specification-version>
                     </run-dependency>
                 </dependency>
+                <dependency>
+                    <code-name-base>org.netbeans.api.lsp</code-name-base>
+                    <build-prerequisite/>
+                    <compile-dependency/>
+                    <run-dependency>
+                        <release-version>1</release-version>
+                        <specification-version>1.8</specification-version>
```
Comment (ppisl) [inline java/refactoring.java/nbproject/project.xml:61]: The new version of LSP API is 1.9. Version 1.8 introduced StructureElement, which is changed in this new version 1.9.

File: java/java.sourceui/src/org/netbeans/modules/java/source/ui/LspElementUtils.java [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        StructureProvider.Builder builder = StructureProvider.newBuilder(createName(info, el), ElementHeaders.javaKind2Structure(el));
+        builder.detail(createDetail(info, el));
+        setOffsets(info, el, builder);
+        if (info.getElements().isDeprecated(el)) {
+            builder.addTag(StructureElement.Tag.Deprecated);
+        }
+
+        if (childAcceptor != null) {
+            for (Element child : el.getEnclosedElements()) {
+                TreePath p = info.getTrees().getPath(el);
```
Comment (ppisl) [inline java/java.sourceui/src/org/netbeans/modules/java/source/ui/LspElementUtils.java:82]: `el` is not changed in this `for` cycle, so why we get `TreePath` in every cycle? Probably it should be `child` the param of `getPath()`?

**Review by ppisl — APPROVED** [review]

File: ide/api.lsp/src/org/netbeans/api/lsp/CallHierarchyEntry.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        public List<CallRange> getRanges() {
+            return ranges;
+        }
+    }
+    
+    /**
+     * Represents a textual range. Range covers text from `startOffset' to `endOffset' (exclusive).
+     */
+    public static final class CallRange {
+        private final int startOffset;
```
[the remark this answers is not in the document]
Comment (sdedic) [reply ide/api.lsp/src/org/netbeans/api/lsp/CallHierarchyEntry.java:111]: OK, will introduce a toplevel `Range` ... plus I guess I could add `getSelectionRange` and `getExtendedRange` to the `StructureElement` - right ?

**Review by dbalek — APPROVED** [review]

**Comment by sdedic** [comment]
Sorry, errors found during testing...

**Pull request merged**